### PR TITLE
Support selective MXFP8 inference precision

### PR DESCRIPTION
**Pull request by wdykas**
Select BF16/MXFP8 inference parameter storage using the same TE per-module precision recipe as training. Routed expert FC1/FC2 modules can use MXFP8 while attention, shared experts, MTP, and the first two/last four layers remain BF16.

The recipe is applied during module construction. BF16 modules allocate BF16 parameters directly, so their checkpoint values never pass through MXFP8 storage and NeMo RL needs no pre-checkpoint callback. With global MXFP8 parameter storage, matching MXFP8 recipes that omit `fp8_param` automatically inherit the enclosing initialization context and its BF16 boundary-layer policy. Explicit `fp8_param: false` opts out.

- Support mixed BF16/MXFP8 expert layers with Torch, FlashInfer, and vLLM-selected inference. The integrated vLLM kernel is BF16-only: its MXFP8 layers fall back to MCore scaled grouped GEMM, not a native vLLM MXFP8 kernel.
- Pass GPT decoder/MTP module names during construction so recipes can choose parameter storage before checkpoint load. This restores only the name propagation removed by main's broad revert #7326, not the other reverted features. Unnamed custom TransformerBlock layer specs receive no new keyword argument.
- Use canonical MCore MXFP8 storage/refit buffers for all three selections. FlashInfer derives its kernel-specific shuffled weights from those buffers and refreshes them after refit.
- Convert only existing MXFP8 parameters to inference storage; preserve BF16 parameter objects and values.
- Reject mixed FC1/FC2 or local-expert precision within one MoE layer before conversion mutates storage.
- Preserve refit names, public parameter metadata, and persistent buffers across discrete and single-grouped TE parameter formats.
- Use separate activation and MXFP8 quantization for SwiGLU; the existing fused quantization kernel supports squared ReLU only.

Split 1 of #6933. Batch-invariant execution and the TE inference backend are outside this PR.

## Validation

After merging upstream main `d046b0de593b1802c78a7664765174c031eb452c`:

- 69 focused tests passed on each of four GB200 ranks: inference configuration, per-module TE initialization, BF16/MXFP8 selection, parameter metadata, grouped refit, and all three selective-MXFP8 M2N backend cases. Complete Hybrid config comparison also passed.
- After the GPT naming fix, the real GPT constructor storage regression passed for both TE training and inference-optimized specs, along with the existing TE recipe test (3 cases, one GB200 rank). It asserts only middle-layer routed experts have MXFP8 parameters before any checkpoint load.
- Black, isort, Ruff, and PR diff whitespace checks passed. Full local Sphinx HTML build with warnings treated as errors passed after documentation cleanup. The post-merge CI docs, lint, and pip/uv installation checks passed at `2e2d9646`; CI has been requested again for GPT naming fix `8faece7a7`.
- After correcting GPT name propagation, Qwen3-30B-A3B completed two NeMo RL GRPO steps using the vLLM-selected backend, including refit, generation, policy scoring, backward/optimizer execution, and the second rollout. The earlier TE warnings about quantized weights without quantized compute are absent in this run. Generation KL errors were approximately 0.0029 and 0.0018, not zero; rewards/losses were zero.

The NeMo RL smoke test uses four-GB200 colocation with TP=1/EP=4, `fp8_param=true`, the routed-expert-only TE recipe, and first-two/last-four BF16 layers. It uses NeMo RL integration revision `e7e36deeeed1b45940a4c788932b5feb2e44fa7f` (PR #3208-based), PyTorch `2.11.0+cu130`, and TE `2.18.0+27486e03`. No additional tracked NeMo RL source changes were needed; this is not a claim of compatibility with unmodified NeMo RL main.

This is a small compatibility smoke test: four responses per step and eight generated tokens. It is not a learning or performance benchmark, and does not target zero training/inference log-probability difference. Batch invariance and CUDA graphs are disabled. Qwen does not exercise shared experts or MTP, and the run does not validate different training/inference parallel layouts.

Nano with FlashInfer completed generation and backward but ran out of GPU memory creating Adam state. A cross-layout retry (training TP=2/ETP=2/EP=2, inference TP=1/EP=4) also completed refit, generation, scoring, and backward, then hit the same Adam-state memory limit. Thus cross-layout execution was exercised, but Nano's optimizer step and second refit/rollout cycle are not validated.

Retrying Nano with CPU optimizer offload instead hit recursive TE MXFP8 CPU/GPU dequantization during `HybridDeviceOptimizer` construction. The CPU/pin-memory failure also reproduces with a tiny TE-only MXFP8 tensor, without importing MCore; a plain BF16 control passes. The failing optimizer line is unchanged from main. Completing Nano training on this setup needs more GPU memory or a separate optimizer CPU-offload fix; optimizer changes are outside this PR.

The earlier selective-MXFP8 M2N CI failure has not reproduced locally, including the preceding refit test sequence. GPU CI confirmation is still required; local passes do not establish that failure is fixed. The exact CI container could not be retrieved with the node's ECR credentials.

### CODE REVIEW

File: megatron/core/transformer/moe/experts.py [hunk 1]
```diff
@@ -1508,9 +1511,16 @@ def forward(
                 permuted_local_hidden_states, permuted_probs, routing_map=routing_map
             )
         elif self.inference_grouped_gemm_backend == InferenceGroupedGemmBackend.VLLM:
+            if isinstance(self._fc1_weight, MXFP8Tensor):
```
Comment (wdykas) [inline megatron/core/transformer/moe/experts.py]: this is like a bit hardcoded I gotta fix this

File: megatron/core/transformer/transformer_config.py [hunk 2]
```diff
@@ -1784,7 +1801,44 @@ def __post_init__(self):
...
+                    raise ValueError(
+                        f"Invalid MXFP8 parameter regex {pattern!r}: {error}"
+                    ) from error
+
         if self.transformer_impl == "inference_optimized" and self.num_moe_experts is not None:
+            try:
+                self.inference_grouped_gemm_backend = InferenceGroupedGemmBackend(
+                    self.inference_grouped_gemm_backend
+                )
+            except ValueError:
```
Comment (wdykas) [inline megatron/core/transformer/transformer_config.py]: move exception in the function

File: megatron/core/parameter_metadata.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+
+"""Shared parameter metadata used by sharding and parameter replacement paths."""
+
+# Metadata that must survive replacing a parameter or exposing member views of a
+# grouped parameter. Refit and tensor-parallel planning inspect these attributes.
+PARAMETER_SHARDING_ATTRIBUTES = (
```
Comment (wdykas) [inline megatron/core/parameter_metadata.py]: look into way to do this without hardcoding the attributes

**Comment by wdykas** [comment]
/ok to test [978a029](https://github.com/NVIDIA/Megatron-LM/pull/7300/commits/978a029b8ca701eea950878a1a2bebe5babc24cd)

File: megatron/core/transformer/moe/experts.py [hunk 4]
```diff
@@ -1322,11 +1323,20 @@ def refresh_flashinfer_mxfp8_weights(self) -> bool:
             or self.inference_grouped_gemm_backend != InferenceGroupedGemmBackend.FLASHINFER
         ):
             return False
+
+        fc1_is_mxfp8 = isinstance(self._fc1_weight, FlashInferRoutedMXFP8Weight)
```
Comment (wdykas) [inline megatron/core/transformer/moe/experts.py]: why is this hardcoded?

File: megatron/core/transformer/moe/experts.py (hunk 1 above)
Comment (wdykas) [reply megatron/core/transformer/moe/experts.py]: Fixed in 178d764f8. The layer now determines precision once from every FC1/FC2 expert weight during lazy construction, records that layer-level decision, and dispatches vLLM from the recorded state. A partial FC1/FC2 selection now fails early with a precise error.

File: megatron/core/transformer/transformer_config.py (hunk 2 above)
Comment (wdykas) [reply megatron/core/transformer/transformer_config.py]: Fixed in 178d764f8. Config conversion and error reporting now live in InferenceGroupedGemmBackend.from_config(); the valid choices are derived from the enum and the original ValueError is preserved as the cause.

File: megatron/core/parameter_metadata.py (hunk 3 above)
Comment (wdykas) [reply megatron/core/parameter_metadata.py]: Fixed in 178d764f8. The attribute allowlist is replaced by copy_parameter_metadata(), which copies public dynamically attached Megatron metadata while excluding private tensor-subclass storage/quantizer state. The new unit test includes an arbitrary future metadata name to ensure this remains allowlist-free.

File: megatron/core/transformer/moe/experts.py (hunk 4 above)
Comment (wdykas) [reply megatron/core/transformer/moe/experts.py]: Fixed in 178d764f8. FlashInfer forward and post-refit refresh now use the precision recorded when the layer constructs its expert stacks, rather than hardcoded FC1/FC2 type checks.

**Comment by wdykas** [comment]
Addressed the latest review in 90555032b57150abb99f3969bc28a7fe58f99a4d:

- both initial conversion and refit conversion now preflight every local FC1/FC2 expert weight and reject a mixed-precision policy before mutating storage; the first-forward guard remains as a backstop
- the rejection tests cover real `weight0` naming, per-projection asymmetry, and atomic failure
- the TE mixed-precision guide now documents `fp8_param: null` inheritance and uses it in the examples
- the inference guide documents the include/exclude fields, FC1/FC2 pairing constraint, all three supported grouped-GEMM choices, vLLM dispatch behavior, and selective refit
- `copy_parameter_metadata` now lives in `megatron/core/utils.py`; `_materialize_mxfp8_parameter_as_bf16` explains why filtered TE storage must become an ordinary BF16 parameter

Validation on 4x GB200: 225 passed / 1 skipped per rank across the affected MXFP8 inference and refit suites; focused review regressions 4 passed; ruff clean; pylint 10.00/10; isort/Black/diff checks clean.

File: megatron/core/extensions/transformer_engine.py [hunk 5]
```diff
@@ -196,6 +197,14 @@ def parse_from_config(cls, quant_config: Dict[Any, Any]) -> "TEQuantizationRecip
...
         Parse config from quantization dictionary.
         """
+        if "fp8_quantization_recipe" in quant_config and "fp8_param" not in quant_config:
+            warnings.warn(
+                "A Transformer Engine per-module FP8 recipe without 'fp8_param' now "
+                "inherits the enclosing parameter-storage context. Set fp8_param=False "
+                "explicitly to preserve the previous forced-BF16 behavior.",
+                UserWarning,
+                stacklevel=2,
+            )
```
Comment (santhnm2) [inline megatron/core/extensions/transformer_engine.py]: This seems risky, should we instead control inheriting the enclosing model-init context through an explicit flag?

File: megatron/core/inference/quantization/utils.py [hunk 6]
```diff
@@ -78,35 +80,154 @@ def resolve_mxfp8_backend(
...
+    ``fp8_param=True`` wraps eligible weights in TE MXFP8 storage during model
+    construction, before the inference include/exclude filters are applied. A
+    filtered-out weight must therefore be dequantized so the BF16 grouped-GEMM
+    and refit paths do not receive TE's quantized tensor subclass. Rebuilding
+    the parameter also requires copying Megatron's public sharding and refit
+    metadata so checkpoint loading and reshard planning retain the same layout.
+    """
+    bf16_parameter = torch.nn.Parameter(
+        parameter.dequantize().to(torch.bfloat16), requires_grad=parameter.requires_grad
+    )
```
Comment (santhnm2) [inline megatron/core/inference/quantization/utils.py]: To confirm, is this basically doing bf16 tensor on disk -> quantize to TEmxfp8 tensor -> dequantize back to bf16 tensor? I guess there's no way to just skip the intermediate quantize because that would require TE to be able to filter internally? How does training handle fine-grained quantization recipes?

File: megatron/core/extensions/transformer_engine.py (hunk 5 above)
Comment (wdykas) [reply megatron/core/extensions/transformer_engine.py]: yeah let me try that now

File: megatron/core/inference/quantization/utils.py (hunk 6 above)
Comment (wdykas) [reply megatron/core/inference/quantization/utils.py]: let me see if I can filter before checkpoint load

Comment (wdykas) [reply megatron/core/inference/quantization/utils.py]: ok I think I can maybe filter before. I just need to be careful in nemo rl

File: megatron/core/extensions/transformer_engine.py (hunk 5 above)
Comment (wdykas) [reply megatron/core/extensions/transformer_engine.py]: made it a flag

**Comment by wdykas** [comment]
/ok to test [b95ef6b](https://github.com/NVIDIA/Megatron-LM/pull/7300/commits/b95ef6b87d2507731d82d7665256677ecb422012)

File: megatron/core/inference/quantization/utils.py (hunk 6 above)
Comment (wdykas) [reply megatron/core/inference/quantization/utils.py]: removed it

**Review by yaoyu-33 — APPROVED** [review]

**Review by zhongbozhu — APPROVED** [review]

File: megatron/core/utils.py [hunk 7]
```diff
@@ -859,6 +859,22 @@ def safely_set_viewless_tensor_data(tensor, new_data_tensor):
     tensor.data = new_data_tensor
 
 
+def copy_parameter_metadata(destination: torch.Tensor, source: torch.Tensor) -> None:
```
Comment (wdykas) [inline megatron/core/utils.py:862]: this is weird but haven't figured out around it

File: megatron/core/extensions/transformer_engine.py [hunk 8]
```diff
@@ -165,9 +165,17 @@ class TEQuantizationRecipe:
...
     """
-    fp8_param: bool = False
+    inherit_model_init_context: bool = False
     """
-    If cast the initialized parameters to fp8 precision and all-gather weights in FP8.
+    Whether parameter storage should always inherit the enclosing model-init context.
+    Matching global and per-module MXFP8 policies also inherit automatically when
+    ``fp8_param`` is omitted.
+    """
+    fp8_param: bool | None = None
```
Comment (santhnm2) [inline megatron/core/extensions/transformer_engine.py]: Can we revert `fp8_param` back to a non-optional bool now that we have `inherit_model_init_context`?

File: megatron/core/transformer/transformer_block.py [hunk 9]
```diff
@@ -362,12 +364,21 @@ def build_layer(layer_spec, layer_number):
...
 
             with quantization_context:
+                # Resolve per-module storage before TE allocates parameters, not
+                # only in GPTModel's post-construction finish_init pass. Avoid
+                # adding kwargs to custom layer specs when naming is unused.
+                layer_kwargs = (
+                    {"name": f"{self.name}.layers.{layer_number - 1}"}
+                    if self.name is not None
+                    else {}
+                )
```
Comment (santhnm2) [inline megatron/core/transformer/transformer_block.py]: @wdykas could you explain why this code is necessary?

Comment (wdykas) [reply megatron/core/transformer/transformer_block.py]: discussed offline will clarify comments for next person

File: megatron/core/extensions/transformer_engine.py (hunk 8 above)
Comment (wdykas) [reply megatron/core/extensions/transformer_engine.py]: We still need to distinguish omitted from explicit false. omitted automatically inherits a matching global MXFP8 storage policy, but false keeps BF16 storage. We could revert to a bool, but then users would need to set inherit_model_init_context explicitly.

**Review by yaox12 — APPROVED** [review]

**Review by santhnm2 — APPROVED** [review]

**Review by kvareddy — APPROVED** [review]

**Comment by wdykas** [comment]
/ok to test [722437c](https://github.com/NVIDIA/Megatron-LM/pull/7300/commits/722437c8af667221a7365375bdeec86040f15cd5)

File: tests/test_utils/recipes/gb200/unit-tests.yaml [hunk 10]
```diff
@@ -46,6 +46,15 @@ spec:
...
 
+    # Keep the M2N transport override in a separate process because NCCL caches
+    # its environment settings. Other GB200 tests retain their default backend.
+    if [[ "$BUCKET" == "tests/unit_tests/resharding/test_nccl_m2n_copy_service.py" ]]; then
+        export NCCL_GIN_TYPE=2
+        export NCCL_NET_PLUGIN=none
+        export NCCL_GIN_PLUGIN=none
+        export NCCL_RMA_PLUGIN=none
+    fi
+
```
Comment (ko3n1g) [inline tests/test_utils/recipes/gb200/unit-tests.yaml]: can we move this into unit test fixtures that set/unset around the tests?

**Review by ko3n1g — CHANGES_REQUESTED** [review]

File: tests/test_utils/recipes/gb200/unit-tests.yaml (hunk 10 above)
Comment (wdykas) [reply tests/test_utils/recipes/gb200/unit-tests.yaml]: I think I changed properly?

**Review by ko3n1g — APPROVED** [review]

**Comment by wdykas** [comment]
/ok to test [3b8c68f](https://github.com/NVIDIA/Megatron-LM/pull/7300/commits/3b8c68f9712b7be6e4abadda959641bdefe4fe95)

**Comment by chtruong814** [comment]
CI ran as part of a batch of PRs in the merge queue to consolidate CI runs
https://github.com/NVIDIA/Megatron-LM/pull/7462
https://github.com/NVIDIA/Megatron-LM/actions/runs/35306160723/job/105485956036

Tests passed, so will fast merge this.
